store/postgres/src/entities.rs

```diff
@@ -0,0 +1,993 @@
+
+    // Note that we have to use conn.batch_execute to issue DDL commands; using
+    // diesel::sql_query(..).execute(conn) can lead to cases where the command
+    // is sent to the database, and shows up in the database logs, but has no
+    // effect at all, but also does not result in an error.
+    let query = format!("create schema {}", schema_name);
+    conn.batch_execute(&*query)?;
+
+    // The order of columns in the primary key matters a lot, since
+    // we want the pk index to also support queries that do not have an id,
+    // just an entity (like counting the number of entities of a certain type)
+    let query = format!(
+        "create table {}.entities(
+            entity       varchar not null,
+            id           varchar not null,
+            data         jsonb,
+            event_source varchar not null,
+            primary key(entity, id)
+        )",
+        schema_name
+    );
+    conn.batch_execute(&*query)?;
+
+    // TODO: should we check that event_source is not "none"?
+    let query = format!(
+        "create trigger entity_change_insert_trigger
+            after insert on {schema}.entities
+            for each row
+            execute procedure subgraph_log_entity_event()",
+        schema = schema_name
+    );
+    conn.batch_execute(&*query)?;
+
+    let query = format!(
+        "create trigger entity_change_update_trigger
+            after update on {schema}.entities
+            for each row
+            when (old.data != new.data)
+            execute procedure subgraph_log_entity_event()",
+        schema = schema_name
+    );
+    conn.batch_execute(&*query)?;
+
+    let query = format!(
+        "create trigger entity_change_delete_trigger
+            after delete on {schema}.entities
+            for each row
+            execute procedure subgraph_log_entity_event()",
+        schema = schema_name
+    );
+    conn.batch_execute(&*query)?;
+
+    let query = format!(
+        "create table {}.entity_history (
+            id           serial primary key,
+	        event_id     integer references event_meta_data(id)
+                             on update cascade on delete cascade,
+            entity       varchar not null,
+	        entity_id    varchar not null,
+	        data_before  jsonb,
+	        reversion    bool not null default false,
+	        op_id        int2 NOT NULL)",
+        schema_name
+    );
+    conn.batch_execute(&*query)?;
+
+    let query = format!(
+        "create index entity_history_event_id_btree_idx
+            on {}.entity_history(event_id)",
+        schema_name
+    );
+    conn.batch_execute(&*query)?;
+
+    Ok(())
+}
+
+#[allow(dead_code)]
+pub(crate) fn drop_schema(
+    conn: &diesel::pg::PgConnection,
+    subgraph: &SubgraphDeploymentId,
+) -> Result<usize, StoreError> {
+    let info = find_schema(conn, subgraph)?;
+    if let Some(schema) = info {
+        let query = format!("drop schema if exists {} cascade", schema.name);
```

Comment:
Will cascade delete triggers as well?

Reply:
Yes, the triggers are attached to the `entities` table, and they will get removed when the table is dropped (of course, it won't get rid of the actual stored procedure that the triggers run)